lib/tools/adb-commands.js

```diff
@@ -262,7 +265,207 @@ methods.push = async function (localPath, remotePath) {
+        }
+      }
+    }
+    return pids;
+  } catch (e) {
+    log.errorAndThrow(`Unable to get pids for ${name}. Orginial error: ${e.message}`);
+  }
+};
+
+methods.killProcessesByName = async function (name) {
+  try {
+    log.debug(`Attempting to kill all ${name} processes`);
+    let pids = await this.getPIDsByName(name);
+    if (pids.length < 1) {
+      log.errorAndThrow(`No ${name} process found to kill`);
+    }
+    for (let pid of pids) {
+      await this.killProcessByPID(pid);
+    }
+  } catch (e) {
+    log.errorAndThrow(`Unable to kill ${name} processes. Original error: ${e.message}`);
+  }
+};
+
+methods.killProcessByPID = async function (pid) {
+  log.debug(`Attempting to kill process ${pid}`);
+  return this.shell(['kill', pid]);
+};
+
+methods.broadcastProcessEnd = async function (intent, processName) {
+  // start the broadcast without waiting for it to finish.
+  this.broadcast(intent);
+  // wait for the process to end
+  let start = Date.now();
+  let timeoutMs = 40000;
+  try {
+    while ((Date.now() - start) < timeoutMs) {
+      if (await this.processExists(processName)) {
+        // cool down
+        await sleep(400);
+        continue;
+      }
+      return;
+    }
+    log.errorAndThrow(`Process never died within ${timeoutMs} ms`);
+  } catch (e) {
+    log.errorAndThrow(`Unable to broadcast process end. Original error: ${e.message}`);
+  }
+};
+
+methods.broadcast = async function (intent) {
+  if (!this.isValidClass(intent)) {
+    log.errorAndThrow(`Invalid intent ${intent}`);
+  }
+  log.debug(`Broadcasting: ${intent}`);
+  await this.shell(['am', 'broadcast', '-a', intent]);
+};
+
+methods.endAndroidCoverage = function () {
+  if (this.instrumentProc) {
+    this.instrumentProc.kill();
+  }
+};
+
+methods.instrument = async function (pkg, activity, instrumentWith) {
+  if (activity[0] !== ".") {
+    pkg = "";
+  }
+  let pkgActivity = (pkg + activity).replace(/\.+/g, '.'); // Fix pkg..activity error
+  let stdout = await this.shell(['am', 'instrument', '-e', 'main_activity',
+                                 pkgActivity, instrumentWith]);
+  if (stdout.indexOf("Exception") !== -1) {
+    log.errorAndThrow(`Unknown exception during instrumentation. ` +
+                      `Original error ${stdout.split("\n")[0]}`);
+  }
+};
+
+methods.androidCoverage = async function (instrumentClass, waitPkg, waitActivity) {
+  if (!this.isValidClass(instrumentClass)) {
+    log.errorAndThrow(`Invalid class ${instrumentClass}`);
+  }
+  return new Promise(async (resolve, reject) => {
+    let args = this.adb.defaultArgs
+      .concat(['shell', 'am', 'instrument', '-e', 'coverage', 'true', '-w'])
+      .concat([instrumentClass]);
+    log.debug(`Collecting coverage data with: ${[this.adb.path].concat(args).join(' ')}`);
+    try {
+      // am instrument runs for the life of the app process.
+      this.instrumentProc = child_process.spawn(this.adb.path, args);
+      this.instrumentProc.on('error', function (e) {
```

Comment:
if you use `SubProcess` this will be easier:

```js
this.instrumentProc = new SubProcess(this.adb.path, args);
await this.instrumentProc.start(0); // this will take care of any errors so you don't have to listen to that event
this.instrumentProc.on('output', (stdout, stderr) => {
  if (stderr) {
    reject(new Error(`Failed to run...`));
  }
});
await this.waitForActivity(waitPkg, waitActivity);
resolve();
```

of course you'll need to then use `await this.instrumentsProc.stop();` instead of `this.instrumentsProc.kill()` elsewhere in the code